[DR-1970] Write directory dao equivalent for Azure

samanehsan opened the pull request.

samanehsan committed:
WIP Handle file directory metadata in azure

samanehsan committed:
Refactor reused firestore code

samanehsan committed:
More refactoring

samanehsan committed:
Formatting fixes

samanehsan committed:
Set directory entity row key

samanehsan committed:
Add unit tests

samanehsan committed:
Fix tabledirectorydao test mocking

giventocode commented on src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java:
```diff
@@ -0,0 +1,241 @@
+public class TableDirectoryDao {
+  private final Logger logger = LoggerFactory.getLogger(FireStoreDirectoryDao.class);
+  private static final String TABLE_NAME = "dataset";
+  private static final String PARTITION_KEY = "partitionKey";
```
Is there a domain/functional value that can be used as a partition key - e.g., filesystem name,  root directory, parent directory, etc.? I understand the tendency towards simplicity and therefore start with a fixed partition.  But it is always better to think through this problem at the beginning vs. later and dealing with unpartitioned data.

samanehsan replied:
That's a good point! Each dataset/snapshot will have a separate storage account with its own tables, so partitioning this table by dataset id won't make a difference. The root directory is the same for all of the files, so I think using the parent directory would work since we can derive that from the file path when writing/retrieving rows.

giventocode replied:
Sounds good. Does it make sense to prefix the parent directory with the dataset id?  I think this will make it easier to understand the data and guarantee uniqueness without relying on the assumption of one storage account per dataset - which could be revisited in the future.

giventocode commented on src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java:
```diff
@@ -0,0 +1,241 @@
+      String rowKey = fileMetadataUtils.encodePathAsFirestoreDocumentName(testPath);
+      TableEntity entity =
+          FireStoreDirectoryEntry.toTableEntity(PARTITION_KEY, rowKey, dirToCreate);
+      TableTransactionAction t =
```
No sure if there's value in using a transaction for a single entity/row. Furthermore, given the fixed partition key, this transaction will lock the entire table.

samanehsan replied:
We create a variable number of entities here so I put it into a single transaction. Specifically, we create an entity for the `createEntity` (corresponding to the file that is being ingested), and for any directories in the file path that do not have a record. 

Also that's a good point about this locking the entire table. What is the behavior for a simultaneous request that tries to read/write to the table? Does it retry depending on the client configuration you mentioned above?

giventocode replied:
Okay, with a variable number of entities makes sense. Transactions are scoped at the partition level. So if you have a single partition, you will lock the whole table and serialize the write operations, with the potential impact of affecting the throughput of the system.  Using a correct partition scheme should help with this as well.

samanehsan replied:
These entities all have different partition keys now, so I removed the transaction here

giventocode commented on src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java:
```diff
@@ -0,0 +1,241 @@
+              .setFilter(
+                  String.format(
+                      "path eq '%s'", fileMetadataUtils.makePathFromLookupPath(lookupPath)));
+      PagedIterable<TableEntity> entities = tableClient.listEntities(options, null, null);
```
Here it seems like we are scanning the whole table. Again, having a partition key to scope/limit the scan is preferred.

nmalfroy replied:
I wonder if there are perf numbers on a scan vs. partition size?

giventocode commented on src/main/java/bio/terra/service/filedata/azure/tables/AzureTableUtils.java:
```diff
@@ -0,0 +1,66 @@
+        .forEach(
+            entity -> {
+              try {
+                retryOperation(entity, generator);
```
The SDK provides retriable capabilities out of the box. The client builder allows you to set an appropriate policy and retry options. More details [here](https://docs.microsoft.com/en-us/java/api/com.azure.data.tables.tableserviceclientbuilder.retrypolicy?view=azure-java-stable#com_azure_data_tables_TableServiceClientBuilder_retryPolicy_com_azure_core_http_policy_RetryPolicy_). 

However, having another retry policy is okay if there is a need to retry due to an expected functional behavior no covered by the SDK. In such a case, a retry policy must be triggered by well-defined exceptions that identify the condition. Ultimately, the goal is to avoid duplicate retries attempts.

samanehsan replied:
Thanks! That's good to know

giventocode commented on src/main/java/bio/terra/service/filedata/azure/tables/TableFileDao.java:
```diff
@@ -96,53 +91,11 @@ public FireStoreFile retrieveFileMetadata(TableServiceClient tableServiceClient,
       throws InterruptedException {
     TableClient tableClient = tableServiceClient.getTableClient(TABLE_NAME);
-    scanTableObjects(
+    azureTableUtils.scanTableObjects(
```
This API uses futures to perform async operations. The SDK also supports reactive programming as an FYI, and I'm bringing this up just in case the async model is flexible for the project. So you can rewrite this with a "subscription" to the deleted event. This approach simplifies the implementation and removes the need of having to wire up an executor. More information here: [Asynchronous programming in the Azure SDK for Java | Microsoft Docs](https://docs.microsoft.com/en-us/azure/developer/java/sdk/async-programming)

samanehsan replied:
That's really cool!

samanehsan replied:
For now I added retry options to the TableServiceClient and BlobContainerClient (and removed the `retryOperation` method, but I do want to revisit this model in a future PR

samanehsan committed:
Fix creating directory entries

samanehsan committed:
More refactoring

samanehsan committed:
Use datasetid + parent dir as partition key

samanehsan committed:
Spotless fixes

samanehsan committed:
Use azure built-in retries

samanehsan committed:
Fix spotbugs errors

tlangs commented on src/main/java/bio/terra/service/filedata/FileMetadataUtils.java:
```diff
@@ -0,0 +1,86 @@
+  public FileMetadataUtils() {}
+
+  public String getDirectoryPath(String path) {
+    String[] pathParts = StringUtils.split(path, '/');
```
Why not use `Paths.get`. `Paths` provides a nice API to get file names, parent directories, etc.

tlangs commented on src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java:
```diff
@@ -0,0 +1,248 @@
+  }
+
+  public String encodePathAsAzureRowKey(String path) {
+    return StringUtils.replaceChars(path, '/', ' ');
```
Why not `path.replaceAll("/", " ")`?

tlangs commented on src/main/java/bio/terra/service/filedata/azure/util/BlobContainerClientFactory.java:
```diff
@@ -22,6 +24,11 @@
   public static final Duration DELEGATED_KEY_DURATION = Duration.ofHours(24);
   private final HttpClient httpClient = HttpClient.createDefault();
+  private static final int MAX_RETRIES = 3;
+  private static final int RETRY_TIMEOUT_SECONDS = 3600;
```
Should these be configurable by the environment, or are these well-known working values?

samanehsan replied:
Yes! I just picked these to start so we'll probably need to adjust them as we test things out

tlangs commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java:
```diff
@@ -512,10 +482,10 @@ private void storeTopDirectory(Firestore firestore, String snapshotId, String di
       for (String testPath = lookupDirPath;
-          !testPath.isEmpty() && !StringUtils.equals(testPath, ROOT_DIR_NAME);
-          testPath = fireStoreUtils.getDirectoryPath(testPath)) {
+          !testPath.isEmpty() && !StringUtils.equals(testPath, FileMetadataUtils.ROOT_DIR_NAME);
```
`"testPath".equals(...)`?

tlangs commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryEntry.java:
```diff
@@ -156,4 +158,65 @@ public String toString() {
+
+  public static FireStoreDirectoryEntry fromTableEntity(TableEntity entity) {
+    return new FireStoreDirectoryEntry()
+        .fileId(entity.getProperty("fileId").toString())
```
String property names could be final class fields since they're used in both serialization and deserialization

snf2ye commented on src/main/java/bio/terra/service/filedata/azure/util/BlobContainerClientFactory.java:
```diff
@@ -22,6 +24,11 @@
   private final HttpClient httpClient = HttpClient.createDefault();
+  private static final int MAX_RETRIES = 3;
+  private static final int RETRY_TIMEOUT_SECONDS = 3600;
+  private final RequestRetryOptions retryOptions =
+      new RequestRetryOptions(
+          RetryPolicyType.EXPONENTIAL, MAX_RETRIES, RETRY_TIMEOUT_SECONDS, null, null, null);
```
Nice! So, this is built-in retry functionality?

samanehsan replied:
Yeah! Jesus commented on it here with some more info https://github.com/DataBiosphere/jade-data-repo/pull/1027#discussion_r684401754

snf2ye reviewed (approved):
LGTM!

samanehsan committed:
Use constants for table property names

samanehsan committed:
Make azure retry options configurable

samanehsan committed:
Refactoring

samanehsan commented:
~I addressed your comments @tlangs! Let me know if it looks good~ ☀️

Oops I still have to change the retries defined in the `BlobContainerClientFactory`

samanehsan committed:
Fix azure test

samanehsan committed:
Debugging config changes

nmalfroy commented on src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java:
```diff
@@ -0,0 +1,247 @@
+          new ListEntitiesOptions()
+              .setFilter(
+                  String.format(
+                      "path eq '%s'", fileMetadataUtils.makePathFromLookupPath(lookupPath)));
```
Should we do some sanitation here?  Could we end up with a case where the lookupPath was something like `' or 1 eq 1'` which would cause all of the things to delete?

nmalfroy replied:
Hmm....I think we're pretty safe.  What you can put in the query is pretty limited I think

tlangs replied:
We could do what I've done here (https://github.com/DataBiosphere/jade-data-repo/pull/1013/files#diff-4a47656762443349b2580bcfaddd2b56e5baaa32105229dd150af8ffe890668cR24) and bas64 encode it.

samanehsan replied:
I could see that going into a separate PR that does this for both firestore and storage tables

nmalfroy commented on src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java:
```diff
@@ -0,0 +1,247 @@
+  }
+
+  public List<String> validateRefIds(
+      TableServiceClient tableServiceClient, List<String> refIdArray) {
```
Should we make the fileId fields GUID (UUID)?  in this case this method would take `List<UUID>`

nmalfroy commented on src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java:
```diff
@@ -0,0 +1,247 @@
+  }
+
+  // Returns null if not found - upper layers do any throwing
+  public FireStoreDirectoryEntry retrieveById(
```
Should we rename `FireStoreDirectoryEntry` to just `DirectoryEntry`?

nmalfroy reviewed (commented):
Cool.  I did a pass and it looks great.  I had a couple of questions that may be worth looking at but may be worth doing in a followon

samanehsan committed:
Update validateRefIds to use UUIDs

samanehsan committed:
Revert getDirectoryPath changes

Note the "root" directory here is actually
"/_dr_" not "/"

samanehsan committed:
Suppress spotbugs test error

samanehsan committed:
Configure retries

samanehsan committed:
Fix getting directory path

samanehsan committed:
Configure retryOptions in BlobContainerClientFactory

samanehsan committed:
Fix integration tests

snf2ye reviewed (approved):
I looked over the last few commits - still LGTM!

samanehsan committed:
Make spotbugs happy

tlangs commented on src/main/java/bio/terra/service/filedata/azure/blobstore/AzureBlobStorePdao.java:
```diff
@@ -63,6 +65,16 @@ public AzureBlobStorePdao(
+        this.resourceConfiguration.getRetryTimeoutSeconds(),
+        null,
+        null,
+        null);
```
This is small, but smells weird. Could we provide an overloaded constructor that does this null-passing for us.

samanehsan replied:
Yeah it's a bit confusing...the null parameters here will use the azure-provided defaults for RequestRetryOptions. I'll add that change into https://github.com/DataBiosphere/jade-data-repo/pull/1027 so I can merge this in :)

tlangs reviewed (approved):
Looks good, some minor comments
